### PR TITLE
feat: Support for s3 compatible destinations in batch exports

### LINKED ISSUE #20312: Support for S3 compatible destinations (MinIO) in batch exports
There exists many S3 compatible destinations out there, in particular MinIO. The only thing required to configure the python boto3 S3 client to work with MinIO is passing one additional configuration (and potentially a couple more optional) to the client initialization (for example, this is the first search result: https://gist.github.com/heitorlessa/5b709df96ea6ac5ddc600545c0683d3b).

Some of these configuration parameters would have to be sourced from the user (like `endpoint_url`, or TLS support). But just exposing one (`endpoint_url`) would add support for MinIO destinations for free. Moreover, this would mean we can get rid of the mocking we do in tests that use MinIO (where we mock the S3 client to pass an `endpoint_url` to our dev stack MinIO). Potentially, this could add support for other S3 compatible destinations, but to keep the scope small let's focus on supporting MinIO by making local tests pass without mocking.

TODO:
1. Expose new optional configuration parameters in `S3BatchExportInputs` (`endpoint_url`, `verify` or other variant to indicate TLS support in MinIO).
2. Add frontend fields for the new configuration parameters.
3. Pass new configuration parameters to S3 client initialization if available.
4. Remove mocking from S3 tests, instead passing `endpoint_url` directly to `S3BatchExportInputs` when running locally without S3 variables configured.
5. ???
6. Profit.

### PR DESCRIPTION
**Pull request by tomasfarias**
## Problem

Closes https://github.com/PostHog/posthog/issues/20312, with the caveat TLS verification is not configurable, so the default applies (verification enabled). Eventually, we may expose a checkbox similar to the one used for self-signed postgres certificates to allow not verifying self-signed certificates, but the frontend is being revamped, so I rather wait until that work is done before touching more UI elements.

## Changes

* Adds backend support for `endpoint_url` in S3 batch exports.
* Adds frontend support for `endpoint_url` in S3 batch export edit form.
* Updated tests to pass `endpoint_url` instead of mocking client.

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Updated tests.

And locally created a batch export with `endpoint_url=http://localhost:19000`:

Screenshot 2024-03-13 at 13-51-51 Edit • S3 MinIO test • Batch Exports • PostHog

Which resulted in exports to local MinIO:
Screenshot 2024-03-13 at 13-52-05 MinIO Console

### CODE REVIEW

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py [hunk 1]
```diff
@@ -523,8 +522,8 @@ async def s3_client(bucket_name, s3_key_prefix):
     using a disposable bucket to run these tests or sticking to other tests that use the
     local development MinIO.
     """
-    async with aioboto3.Session().client("s3") as minio_client:
-        yield minio_client
+    async with aioboto3.Session().client("s3") as s3_client:
+        yield s3_client
 
         await delete_all_from_s3(minio_client, bucket_name, key_prefix=s3_key_prefix)
```
Comment (tkaemming) [inline posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:528]: Is this now referencing [the module-scoped `minio_client`](https://github.com/PostHog/posthog/blob/e85168cbafd9f61ba31eefbd2343b25903255627/posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py#L120-L138)? 🤔 I'm confused how that is working…

I tried to run these locally to poke at it myself but these tests are skipped by default so I didn't take the time to investigate further. I'm not sure if these are being run in CI or not, either -- [there are some tests that are skipped from this suite](https://github.com/PostHog/posthog/actions/runs/8266168271/job/22613553157?pr=20894), I suspect that might be these but without verbose output it's hard to say for certain.

**Review by tkaemming — COMMENTED** [review]
The main functional changes look good to me, but I had a question about the tests. I don't feel very qualified to review the UI changes, but if you're confident everything is good there, I'm happy to approve!

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py [hunk 2]
```diff
@@ -480,18 +483,14 @@ async def test_s3_export_workflow_with_minio_bucket(
             workflow_runner=UnsandboxedWorkflowRunner(),
         ):
             # We patch the S3 client to return our client that targets MinIO.
```
[the remark this answers is not in the document]
Comment (tomasfarias) [reply posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:485]: Ah, yes indeed, this one slipped by.

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py (hunk 1 above)
Comment (tomasfarias) [reply posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:527]: Yeah, this should be deleting using the `s3_client`. I missed the name update.

Tests are skipped by default when no S3 credentials are present, there is information in the README on how to run them.

Comment (tomasfarias) [reply posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:527]: We were also not passing the S3 client down when calling `assert_clickhouse_events_in_s3`. Anyways, I've fixed both of these renaming issues, and I've also created a specific key for development testing, as my personal credentials were not working, and it's better to have a dedicated user just for testing. I've Updated the README to reflect this. S3 test is now passing.

**Comment by tomasfarias** [comment]
Had to rebase due to some unrelated storybook tests failing. This should be good to merge once CI is green!

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py [hunk 3]
```diff
@@ -595,20 +593,20 @@ async def test_s3_export_workflow_with_s3_bucket(
             )
 
     workflow_id = str(uuid4())
+    destination_config = s3_batch_export.destination.config | {
```
Comment (tkaemming) [inline posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:596]: Oh cool, I wasn't aware of this `|` syntax — that's useful.

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py (hunk 1 above)
Comment (tkaemming) [reply posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py]: > Tests are skipped by default when no S3 credentials are present, there is information in the README on how to run them, as we do not run them in CI. Doing so would require populating CI with a lot of credentials. Granted, maybe that's something we will want to do in the future...

Hm, yeah… I do worry with these kinds of things that having to intentionally opt-in to running the tests makes them less likely to catch regressions, but this seems okay for now? I'd assume the MinIO targeted tests would likely catch the majority of issues anyway.

**Review by tkaemming — APPROVED** [review]
Looks good to me!

**Comment by tomasfarias** [comment]
> Hm, yeah… I do worry with these kinds of things that having to intentionally opt-in to running the tests makes them less likely to catch regressions, but this seems okay for now? I'd assume the MinIO targeted tests would likely catch the majority of issues anyway.

I agree. I've raised the concern already to see if we can have a better story about managing credentials in CI.

File: posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py (hunk 1 above)
Comment (tomasfarias) [reply posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py]: I agree: opt-in tests means tests that are never ran. I've raised the concern already to see if we can have a better story for managing credentials in CI.

**Pull request merged**